File: backend/app/services/email_service.py
import smtplib
from email.mime.multipart import MIMEMultipart
from email.mime.text import MIMEText
from typing import Dict, Any
import logging
from fastapi import BackgroundTasks
from ..core.config import settings

logger = logging.getLogger(__name__)

class EmailService:
    def __init__(self):
        self.sender_email = settings.SMTP_USER
        self.smtp_server = settings.SMTP_SERVER
        self.smtp_port = settings.SMTP_PORT
        self.smtp_user = settings.SMTP_USER
        self.smtp_password = settings.SMTP_PASSWORD

    def _create_booking_confirmation_email(self, recipient: str, booking_data: Dict[str, Any]) -> MIMEMultipart:
        """Create a booking confirmation email message"""
        message = MIMEMultipart("alternative")
        message["Subject"] = "Passport Appointment Confirmation"
        message["From"] = self.sender_email
        message["To"] = recipient

        # Create HTML content
        html = f"""
        <html>
        <head>
            <style>
                body {{ font-family: Arial, sans-serif; }}
                .container {{ max-width: 600px; margin: 0 auto; padding: 20px; }}
                .header {{ background-color: #004080; color: white; padding: 10px; text-align: center; }}
                .content {{ padding: 20px; }}
                .footer {{ text-align: center; font-size: 12px; color: #666; padding: 10px; }}
                .info {{ margin: 15px 0; }}
                .ref-number {{ font-weight: bold; color: #004080; }}
            </style>
        </head>
        <body>
            <div class="container">
                <div class="header">
                    <h1>Passport Appointment Confirmation</h1>
                </div>
                <div class="content">
                    <p>Dear {booking_data.get('user_name', 'Valued Customer')},</p>
                    <p>Your passport appointment has been successfully booked.</p>

                    <div class="info">
                        <p><strong>Booking Reference:</strong> <span class="ref-number">#{booking_data.get('id', 'N/A')}</span></p>
                        <p><strong>Date:</strong> {booking_data.get('date', 'N/A')}</p>
                        <p><strong>Time:</strong> {booking_data.get('time', 'N/A')}</p>
                        <p><strong>Location:</strong> {booking_data.get('location', 'N/A')}</p>
                    </div>

                    <p>Please arrive 15 minutes before your appointment time with all required documents.</p>
                    <p>If you need to reschedule or cancel, please log in to your account.</p>
                </div>
                <div class="footer">
                    <p>This is an automated message. Please do not reply to this email.</p>
                </div>
            </div>
        </body>
        </html>
        """

        # Attach HTML content
        message.attach(MIMEText(html, "html"))
        return message

    def send_booking_confirmation(self, recipient_email: str, booking_data: Dict[str, Any]) -> bool:
        """
        Send a booking confirmation email to the user
        Returns True if successful, False otherwise
        """
        try:
            message = self._create_booking_confirmation_email(recipient_email, booking_data)

            with smtplib.SMTP(self.smtp_server, self.smtp_port) as server:
                server.starttls()
                server.login(self.smtp_user, self.smtp_password)
                server.send_message(message)

            logger.info(f"Booking confirmation email sent to {recipient_email}")
            return True
        except Exception as e:
            logger.error(f"Failed to send booking confirmation email: {str(e)}")
            return False

    def send_booking_confirmation_async(
        self, background_tasks: BackgroundTasks, recipient_email: str, booking_data: Dict[str, Any]
    ):
        """Add email sending to background tasks"""
        background_tasks.add_task(self.send_booking_confirmation, recipient_email, booking_data)

File: backend/src/services/emailService.ts
import nodemailer from 'nodemailer';
import fs from 'fs';
import path from 'path';
import Handlebars from 'handlebars';
import logger from '../utils/logger';

class EmailService {
  private transporter: nodemailer.Transporter;

  constructor() {
    // For production, use environment variables for these settings
    this.transporter = nodemailer.createTransport({
      host: process.env.EMAIL_HOST || 'smtp.example.com',
      port: Number(process.env.EMAIL_PORT) || 587,
      secure: process.env.EMAIL_SECURE === 'true',
      auth: {
        user: process.env.EMAIL_USER || 'user@example.com',
        pass: process.env.EMAIL_PASS || 'password',
      },
    });
  }

  async sendBookingConfirmation(booking: any, user: any): Promise<void> {
    try {
      const template = this.loadTemplate('booking-confirmation');
      const html = template({
        userName: user.name,
        bookingId: booking.id,
        date: new Date(booking.date).toLocaleDateString(),
        location: booking.locationName || 'Our Location',
        slot: booking.slotName || 'Your Time Slot'
      });

      await this.transporter.sendMail({
        from: process.env.EMAIL_FROM || '"Appointment System" <appointments@example.com>',
        to: user.email,
        subject: 'Your Booking Confirmation',
        html
      });

      logger.info(`Booking confirmation email sent to ${user.email}`);
    } catch (error) {
      logger.error(`Failed to send booking confirmation email: ${error}`);
      // Don't throw error, just log it - don't want email failure to break booking process
    }
  }

  private loadTemplate(templateName: string) {
    try {
      const templatePath = path.join(__dirname, '../templates', `${templateName}.hbs`);
      const templateSource = fs.existsSync(templatePath)
        ? fs.readFileSync(templatePath, 'utf8')
        : '<h1>Booking Confirmation</h1><p>Dear {{userName}},</p><p>Your booking (ID: {{bookingId}}) has been confirmed for {{date}} at {{location}}.</p>';

      return Handlebars.compile(templateSource);
    } catch (error) {
      logger.error(`Failed to load email template: ${error}`);
      // Return a simple backup template
      return Handlebars.compile('<h1>Booking Confirmation</h1><p>Your booking has been confirmed.</p>');
    }
  }
}

export default new EmailService();
